ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/HeartbeatHandler.java

```diff
@@ -0,0 +1,351 @@
+    scheduledExecutorService.scheduleAtFixedRate(discoverTask, 0, configs.discoverClusterIntervalMs,
+                                                 TimeUnit.MILLISECONDS);
+
+  }
+
+  void shutDown() {
+    scheduledExecutorService.shutdown();
+  }
+
+  public ConcurrentHashMap<HostInfo, TreeMap<Long, HeartbeatInfo>> getReceivedHeartbeats() {
+    return receivedHeartbeats;
+  }
+
+  public void registerHeartbeat(HostInfo hostInfo, long timestamp) {
+    receivedHeartbeats.putIfAbsent(hostInfo, new TreeMap<>());
+    TreeMap<Long, HeartbeatInfo> heartbeats = receivedHeartbeats.get(hostInfo);
+
+    synchronized (heartbeats) {
+      heartbeats.put(timestamp, new HeartbeatInfo(timestamp));
+    }
+  }
+
+  ConcurrentHashMap<URI, Boolean> getRemoteHostsURI() {
+    return remoteHostsURI;
+  }
+
+  public Map<String, Boolean> getAliveHosts() {
+    return aliveHosts;
+  }
+
+  class CheckHeartbeatTask implements Runnable {
+
+    @Override
+    public void run() {
+      long now = System.currentTimeMillis();
+      long windowStart = now - configs.heartbeatWindowMs;
+      processHeartbeats(windowStart, now);
+    }
+
+    @VisibleForTesting
+    void runWithWindow(long windowStart, long windowEnd) {
+      processHeartbeats(windowStart, windowEnd);
+    }
+
+    private void processHeartbeats(final long windowStart, final long windowEnd) {
+      if (receivedHeartbeats.isEmpty()) {
+
+        remoteHostsInfo.forEach((host, bool) ->
+                                    aliveHosts.compute(host.toString(), (key, val) -> false));
+      }
+      for (Entry<HostInfo, TreeMap<Long, HeartbeatInfo>> entry : receivedHeartbeats.entrySet()) {
+        HostInfo host = entry.getKey();
+        TreeMap<Long, HeartbeatInfo> heartbeats = entry.getValue();
+        synchronized (heartbeats) {
+          // 1. remove heartbeats older than window
+          while (!heartbeats.isEmpty()) {
+            Entry<Long, HeartbeatInfo> nestedEntry = heartbeats.firstEntry();
+            long timestamp = nestedEntry.getKey();
+            if (timestamp < windowStart) {
+              heartbeats.pollFirstEntry();
+            } else {
+              break;
+            }
+          }
+          // 2. count consecutive missed heartbeats and mark as up
+          downPolicy(host, windowStart, windowEnd, heartbeats);
+        }
+      }
+    }
+
+
+    private void downPolicy(final HostInfo host, final long windowStart, final long windowEnd,
+                            final TreeMap<Long, HeartbeatInfo> heartbeats) {
+      long missedCount = 0;
+      long prev = windowStart;
+      // No heartbeat received in this window
+      if (heartbeats.isEmpty()) {
+        aliveHosts.put(host.toString(), false);
+        return;
+      }
+      for (long ts : heartbeats.keySet()) {
+
+        //Don't count heartbeats after window end
+        if (ts >= windowEnd) {
+          continue;
+        }
+        if (ts - configs.heartbeatSendIntervalMs > prev) {
```

Comment:
Since network is not zero delay, wondering if we would need some padding here? Or may be `"heartbeat.check.interval.ms"` being 2x the send interval is that padding? Just thinking out loud

Reply:
Hmm, you raise a good point. We may need some padding. I opened an issue for it to check once we run it in the wild. #4215